File: zs-leetcode-go/LinkedListCycle/LinkedListCycle.go
package LinkedListCycle

import "for-an-offer/zs-leetcode-go/types"

func hasCycle(head *types.ListNode) bool {
	first, second := head, head.Next
	for second != nil && second.Next != nil && first != second {
		first = first.Next
		second = second.Next.Next
	}
	if first == second && first != nil {
		return true
	}
	return false
}

File: sjq-leetcode-C++/linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if(head==NULL) return false;
        set<ListNode*> s;

        ListNode* temp = head;
        while(temp){
            if(s.find(temp) != s.end())
                return true;
            s.insert(temp);
            temp = temp->next;
        }
        
        return false;
    }
};
